Multicopter auto - Yaw at waypoint before move

bresch opened the pull request:
**Describe problem solved by the proposed pull request**
In some cases, the user wants the drone to yaw first and then move instead of doing a simultaneous yaw and linear motion. A few examples are:
- A drone with a front facing anti-collision sensor (camera, rangefinder, ...)
- A standard vtol plane (low yaw speed and lateral motions are to avoid)
- A highspeed multirotor with specific aerodynamic properties

**Test data / coverage**
SITL tests with `MPC_POS_MODE` 0 and 1.
On the plot below, you can see that the drone starts to move only when the yaw setpoint reaches the desired heading.
![2019-06-27_13-42-20_01_plot](https://user-images.githubusercontent.com/14822839/60266780-11815000-98e9-11e9-9e29-41547c70b4b3.png)


**Usage**
Set `MPC_YAW_MODE = 4`

**Note** that the proposed solution only cares about setpoints and doesn't check for the actual yaw of the vehicle. If the yaw tracking of your vehicle is bad, improve it or reduce `MPC_YAWR_AUTO_MAX`

fixes #11053 

@mrivi Can you try that with avoidance please?

mrivi commented:
@bresch 
This PR 
![pr_yaw](https://user-images.githubusercontent.com/15078736/60268815-91a9b480-98ed-11e9-8731-05814c162b8a.png)

Master

![master_yaw](https://user-images.githubusercontent.com/15078736/60268823-979f9580-98ed-11e9-904c-f5aacde0b00e.png)

bresch commented:
@PX4/testflights Can you try this PR with the parameter `MPC_YAW_MODE = 0,1,2,3 and 4` please?

```
0 towards waypoint
1 towards home
2 away from home
3 along trajectory
4 towards waypoint (yaw first)
```

xdwgood commented:
This parameter is automatically selected only when the autonomous flight (automatic avoidance) is turned on, otherwise the default is selected. Is this PR doing this? :smiley:

Stifael commented:
I have a couple questions:
> In some cases, the user wants the drone to yaw first and then move instead of doing a simultaneous yaw and linear motion

Unfortunately the way the navigator is written, you basically always have to yaw first before proceeding the next waypoint **IF** the vehicle heading is set to point towards the next waypoint. Otherwise the heading check within `mission_block`  (https://github.com/PX4/Firmware/blob/master/src/modules/navigator/mission_block.cpp#L368) is not true and will introduce a sudden brake. I am actually surprised that this has not yet happend with the new smoothing algorithm (I haven't had time yet to test the smoothing yet).

bresch commented on src/lib/FlightTasks/tasks/AutoLine/FlightTaskAutoLine.cpp:
```diff
@@ -49,8 +49,19 @@ void FlightTaskAutoLine::_generateSetpoints()
+		// Wait for the yaw setpoint to be aligned
+		if (!_position_locked) {
+			_velocity_setpoint.setAll(0.f);
+			_position_setpoint = _position;
```
There is no smoothing in that flight task, setting velocity to 0 would lead to some drift.

bresch commented on src/modules/mc_pos_control/mc_pos_control_params.c:
```diff
@@ -750,11 +750,12 @@ PARAM_DEFINE_FLOAT(MPC_SPOOLUP_TIME, 1.0f);
  * @value 1 towards home
  * @value 2 away from home
  * @value 3 along trajectory
+ * @value 4 towards waypoint (yaw first)
```
This is not exactly correct. the linear acceleration was limited if the yaw error was larger than some value. Here, it truly locks the vehicle. Maybe the difference is only noticeable on the drone that has low yaw authority...

bresch commented on src/lib/FlightTasks/tasks/AutoLine/FlightTaskAutoLine.cpp:
```diff
@@ -49,8 +49,19 @@ void FlightTaskAutoLine::_generateSetpoints()
+	if (_param_mpc_yaw_mode.get() == 4 && !_yaw_sp_aligned) {
+		// Wait for the yaw setpoint to be aligned
+		if (!_position_locked) {
+			_velocity_setpoint.setAll(0.f);
```
I wanted to decouple completely yaw from the linear accelerations. This is just a preference.

bresch commented:
@xdwgood I don't think we should couple those because voidance doesn't necessarily mean that the vehicle has a single front facing sensor. 

@Stifael It's only performing that check in navigator if the yaw is defined in a mission item. In out case, the yaw setpoint is generated in the FlightTaskAuto and the Navigator check is not done.

Btw, I added the logic to force yaw before linear acceleration in the standard line tracking algorithm to be consistent but I am not sure is this is really required since it is already limiting the lateral acceleration when the yaw isn't aligned.

dannyfpv commented:
Tested on Pixhawk4 v5 f-450
Modes Tested: 

Position Mode: no issue 
Altitude Mode: no issue 
Stabilized Mode: no issue
mission mode : no issues

pr log: towards waypoint:
https://review.px4.io/plot_app?log=94089c0d-7833-4f81-98b4-0b32f37260d4

pr log: towards home:
https://review.px4.io/plot_app?log=c04830b5-16ba-4b0a-8c67-d260ec7fbe7c

pr log: away from home:
https://review.px4.io/plot_app?log=d85f24b3-992c-4b10-8788-7339ce28f84f

pr log: along trajectory:
https://review.px4.io/plot_app?log=b9a7ab4a-7a69-4698-a580-06e425fbec21

jorge789 commented:
### Tested on PixRacer V4:
Modes Tested

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Automated): Good.
**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**

No issues noted, good flight in general.

**Log:** The MPC_YAW_MODE 0 towards waypoint parameter has changed.
Note:The vehicle flippel when landing by the strong air.
https://review.px4.io/plot_app?log=2f682235-2328-4974-a798-6f4f66cae5e3

**Log** The MPC_YAW_MODE 1 towards home parameter has changed.
https://review.px4.io/plot_app?log=b6257d4d-523c-4c5d-baed-4ac7b769085b

**Log** The MPC_YAW_MODE 2 away from home parameter has changed.
https://review.px4.io/plot_app?log=0f19cc06-e33a-4e6b-8abb-31334d05257a

**Log** The MPC_YAW_MODE 3 along trajectory parameter has changed.
https://review.px4.io/plot_app?log=6acc9c06-4137-4d91-bfac-f8db9de87b02

jorge789 commented:
### Tested on PixRacer V4:

**Flight Card 2**

Modes Tested
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

**Note:**
At the landing the vehicle began to oscillate and descend at the same time, but the descent was not continuous, had brief breaks.Wind 6.0 meters per second

**Log: The MPC_YAW_MODE 0 towards waypoint parameter has changed.**
https://review.px4.io/plot_app?log=50ae8918-64d8-4777-ae15-42a4fca30c65

**Log: The MPC_YAW_MODE 1 towards home parameter has changed.**
https://review.px4.io/plot_app?log=42f20474-bc33-4d8c-9127-3b77d142306e

**Log: The MPC_YAW_MODE 2 away from home parameter has changed.**
https://review.px4.io/plot_app?log=f7f8bcfa-bec5-44ca-b65a-b52c58445f36

**Log: The MPC_YAW_MODE 3 along trajectory parameter has changed.**
https://review.px4.io/plot_app?log=0fae0055-1baa-41c2-8843-f47412974d37

Junkim3DR commented:
### Tested on Pixhawk Pro v4:

**Modes Tested**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL (Automated): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Logs:**

- MPC_YAW_MODE at 0 towards waypoint.
https://review.px4.io/plot_app?log=b093fe86-ffdd-427a-86b8-4a16c47d7294

- MPC_YAW_MODE at 1 towards home.
https://review.px4.io/plot_app?log=0e6c95ea-88c5-4cae-b651-9234d45cc568

- MPC_YAW_MODE 2 away from home. 
https://review.px4.io/plot_app?log=f0a4ee05-a569-4470-bdc5-abb79d110961

- MPC_YAW_MODE at 3 along trajectory.
https://review.px4.io/plot_app?log=5137497a-5214-41b8-941a-bb9398cefe5e

jkflying commented:
Test flew with avoidance with MPC_YAW_MODE=4. The mission at the end is with the avoidance enabled:
Flight log: https://logs.px4.io/plot_app?log=09d60813-d0b3-4f3b-83f0-0459eb5c6bb6

This was flown immediately before with MPC_YAW_MODE=0 and MPC_AUTO_MODE=1, same mission:
https://logs.px4.io/plot_app?log=28337c55-137d-4003-a477-d92dd7e4acc5

It seemed to have strange behavior when turning. It would go back to the location where it was when the new setpoint was received (while yawing), instead of coming to a standstill.
This gave it a wobbly appearance, and it had lots of weird 'overshoot'-like behavior any time there was a yaw as can be seen in the picture.

MPC_YAW_MODE=4
![bokeh_plot(3)](https://user-images.githubusercontent.com/1746276/60715912-727ada80-9f1e-11e9-9dbb-00160036971f.png)

MPC_YAW_MODE=0 and MPC_AUTO_MODE=1
![bokeh_plot(2)](https://user-images.githubusercontent.com/1746276/60715914-727ada80-9f1e-11e9-9406-67096bf4bf02.png)

bresch commented:
@jkflying As you can see in the first 2D plot you shared and the one here below, the vehicle overshoots the setpoint, it's not the setpoint that overshoots
![bokeh_plot(20)](https://user-images.githubusercontent.com/14822839/60883345-41fead80-a24a-11e9-8f02-c7369fa4b29c.png)
For me, the problem is that the velocity/position loops are poorly tuned. The reason why you didn't see it with the other mission mode is that is resets the setpoint to the current position when the vehicle reaches the waypoint. This is not something I would since such a reset produce infinite jerk, acceleration and velocity.

bresch commented:
@jkflying I rebased on top of master, can you review again please such that we can merge?
Btw, the diff in `FlightTaskAutoLineSmoothVel.cpp` looks large, but in fact, I just added : 
```
	if (_param_mpc_yaw_mode.get() == 4 && !_yaw_sp_aligned) {
		// Wait for the yaw setpoint to be aligned
		_velocity_setpoint.setAll(0.f);
        } else {
```
the rest is just indentation

bresch commented:
12 bytes overflow on v2 xD
